### ISSUE TITLE
Destination Filename defaults to subtitle filename when adding multiple files

### ISSUE THREAD

**Issue by 791** [mbunkus/mkvtoolnix#2058]
When adding a movie file and subtitle file with add file dialog (both at once) the destination filename defaults to subtitle filename and subtitle file moves to position 1 in tracks listview. 

The filter should be set to parse the movie title only to fill the destination filename textbox. 
The desired default subtitle file position should be a preference selection.

note - the preference selection for destination filename still inefective.

**Comment by mbunkus** [comment]
> When adding a movie file and subtitle file with add file dialog (both at once) the destination filename defaults to subtitle filename and subtitle file moves to position 1 in tracks listview.

The GUI doesn't impose any order here; it processes the file names in the order they're returned by the file selection dialog. If the subtitle is the first one, then so be it. I could add sorting by file name, but that is what the GUI's file dialog seems to do already — and therefore not change anything.

I will not change where the GUI inserts files. The current way is very predictable: they're inserted at the end of the current layer (if you're "adding", then they're added on the top layer; if you're "appending" they become the last child of the file they're appended to). Making some pieces of information be treated in a special way is a sure-fire way to confuse users.

> The filter should be set to parse the movie title only to fill the destination filename textbox.

I really don't know what you mean by that. This sounds like an enhancement, because at the moment the code for setting the output file name only takes the file names of source files into account but never attributes such as the movie title. Again, a straight-forward mechanisms users can understand without a lot of explanation.

If you want the movie title to be used as the output file name, then that's just a single click of the keyboard away. Just press `Ctrl+T` or select the menu option "Multiplexer" → "Copy file title to destination file name".

I won't add code to set the output file name *automatically* based on the file title.

> note - the preference selection for destination filename still inefective.

And you have to be much more specific that "it doesn't work".

In summary this sounds much like "make the GUI work exactly the way I want it to work", and that's not something I'm willing to do.

**Comment by 791** [comment]
Please don't dismiss my effort as a waste of time here.
When selecting multiple files in add source dialog It ends up with "2_Eng.mkv" etc. (being the subtitle filname with a movie extension added to it. not .srt) Same as if I added the subtitle file first individually.
Obviously it is doing something with the extension - that part needs to be improved.

It would be simpler if the destination file textbox didn't get filled until a movie file is added and then only use that filename ignoring the others. As when adding multiple files at once.

No I haven't poured through the code yet to give example specifics. but I will in time if needed.

**Comment by mbunkus** [comment]
> Obviously it is doing something with the extension - that part needs to be improved.

The extension is chosen based on the selected tracks. If at least one video track is present and selected, `.mkv` is used. If no video track is present but at least one audio track, then `.mka` is used, otherwise `.mks`. If WebM mode is active, then `.webm` is always used regardless of the track types.

The extension is updated every time something relevant changes. If you add a subtitle first, the extension will be `.mks`. If you add a file with a video track later (no matter that it isn't the first file), then the extension is changed to `.mkv`.

This part is working just fine, and I see no need to improve it.

> It would be simpler if the destination file textbox didn't get filled until a movie file is added and then only use that filename ignoring the others. As when adding multiple files at once.

You think so, others don't.

Your arguments don't convince me to spend my time on making your special workflow quicker for you. Sorry.

**Comment by mbunkus** [comment]
During today's workout I've thought a bit about what I could implement with relatively little effort that might help you. I've reconsidered my initial decision and will implement two things:

1. I'll add a new entry to the "files" list box's context menu called "set destination file name from select source file's name". This will switch the GUI's internal state in a way that it thinks that the selected file was the first one added and not whichever other file actually was. Additionally using that option will force an update to the destination file name once. This option can be used in situations such as yours to manually  override what the GUI thinks it should use as the file name template.
2. Additionally I'll add an option in the preferences called "only consider source files with video tracks for automatic file name generation". If that option is enabled, adding a subtitle-only file as the first one won't cause a file name to be set. It'll be off by default so that the current behavior is preserved, but you can simply switch it on and rest assured that your subtitle file won't interfere anymore.

**Comment by mbunkus** [comment]
New [pre-builds for Windows](https://mkvtoolnix.download/windows/pre/) that include both features are available.
